gmail/gmail-connector/gmail-connector-2.0.0/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/GmailIMAPClientLoader.java

```diff
@@ -0,0 +1,131 @@
+import org.apache.synapse.MessageContext;
+import org.apache.synapse.core.axis2.Axis2MessageContext;
+import org.wso2.carbon.connector.core.ConnectException;
+
+import com.google.code.com.sun.mail.imap.IMAPStore;
+import com.google.code.javax.mail.MessagingException;
+
+/**
+ * Class which loads the IMAP store according to the authentication mode.
+ */
+public class GmailIMAPClientLoader {
+
+    /**
+     * Log instance.
+     */
+    private static Log log = LogFactory.getLog(GmailIMAPClientLoader.class);
+
+    /**
+     * Method which loads the IMAPStore instance according to the authentication
+     * mode.
+     *
+     * @param messageContext
+     *            Message context where the instantiated IMAPStore instance is
+     *            stored.
+     * @return
+     *         the loaded IMAPStrore instance.
+     * @throws MessagingException
+     *             as a result of authentication failures
+     * @throws ConnectException
+     *             as a result of invalid configuration
+     */
+    public IMAPStore loadIMAPStore(MessageContext messageContext) throws MessagingException,
+            ConnectException {
+
+        org.apache.axis2.context.MessageContext axis2MsgCtx =
+                ((Axis2MessageContext) messageContext).getAxis2MessageContext();
+        Object prestoredInstance =
+                axis2MsgCtx.getOperationContext()
+                        .getProperty(GmailConstants.GMAIL_IMAP_STORE_INSTANCE);
+
+        // Use if there exists an already stored IMAPStore instance.
+        if (prestoredInstance != null) {
+            log.info("Retriving the prestored IMAPstore instance");
+            return (IMAPStore) prestoredInstance;
+        }
+
+        // Login mode should have been defined during either "init" or
+        // "passwordLogin" operations.
+        Object loginMode = axis2MsgCtx.getProperty(GmailConstants.GMAIL_LOGIN_MODE);
+        if (loginMode == null) {
+            String errorLog = "Gmail configuration details were not initialized";
+            log.error(errorLog);
+            ConnectException connectException = new ConnectException(errorLog);
+            throw (connectException);
+        }
+
+        IMAPStore store = null;
+
+        // Perform SASL authentication if configured using the "Password Login"
+        // operation.
+        if (loginMode.toString().equals(GmailConstants.GMAIL_SASL_LOGIN_MODE)) {
+            log.info("SASL authentication starts");
+            try {
+                store =
+                        GmailSASLAuthenticator.connectToIMAP(messageContext.getProperty(GmailConstants.GMAIL_USER_USERNAME)
+                                        .toString(),
+                                messageContext.getProperty(GmailConstants.GMAIL_USER_PASSWORD)
+                                        .toString());
+            } catch (MessagingException e) {
+                log.error("Failure in SASL authentication");
+                throw (e);
```

Comment:
Any reason for logging and throwing the exception in same block.. Anyway, please log the exception stacktrace as well..